applicationset/generators/git.go

```diff
@@ -222,17 +212,108 @@ func (g *GitGenerator) generateParamsFromGitFile(filePath string, fileContent []
+		err = g.appendExtraParams(git, goTemplateOptions, useGoTemplate, noRevisionCache, verifyCommit, params)
+		if err != nil {
+			return nil, err
+		}
+
 		res = append(res, params)
 	}
 
 	return res, nil
 }
 
+func (g *GitGenerator) appendExtraParams(git *argoprojiov1alpha1.GitGenerator, goTemplateOptions []string, useGoTemplate bool, noRevisionCache bool, verifyCommit bool, params map[string]interface{}) error {
+	extraParams, err := g.getExtraParams(git, goTemplateOptions, useGoTemplate, noRevisionCache, verifyCommit, params)
+	if err != nil {
+		return fmt.Errorf("failed to append extra parameters from files: %w", err)
+	}
+	if len(extraParams) == 0 {
+		return nil
+	}
+	if useGoTemplate {
+		params["extraParams"] = extraParams
+	} else {
+		flat, err := flatten.Flatten(extraParams, "", flatten.DotStyle)
+		if err != nil {
+			return fmt.Errorf("error flattening extra params: %w", err)
+		}
+		for k, v := range flat {
+			params["extraParams."+fmt.Sprintf("%v", k)] = fmt.Sprintf("%v", v)
+		}
+	}
+	return nil
+}
+
+func unMarshalYamlFile(fileContent []byte) ([]map[string]interface{}, error) {
+	objectsFound := []map[string]interface{}{}
+
+	// First, we attempt to parse as an array
+	err := yaml.Unmarshal(fileContent, &objectsFound)
+	if err != nil {
+		// If unable to parse as an array, attempt to parse as a single object
+		singleObj := make(map[string]interface{})
+		err = yaml.Unmarshal(fileContent, &singleObj)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse file: %w", err)
+		}
+		objectsFound = append(objectsFound, singleObj)
+	} else if len(objectsFound) == 0 {
+		// If file is valid but empty, add a default empty item
+		objectsFound = append(objectsFound, map[string]interface{}{})
+	}
+	return objectsFound, nil
+}
+
+func (g *GitGenerator) getExtraParams(git *argoprojiov1alpha1.GitGenerator, goTemplateOptions []string,
+	useGoTemplate, noRevisionCache, verifyCommit bool, params map[string]interface{},
+) (map[string]interface{}, error) {
+	extraParams := map[string]interface{}{}
+	extraParamFilesContent := map[string][]byte{}
+
+	// Keep the order to apply the params in the list of the files provided
+	for _, requestedExtraFilePath := range git.ExtraParameterFiles {
+		templatedPath, err := replaceTemplatedString(requestedExtraFilePath, params, useGoTemplate, goTemplateOptions)
+		if err != nil {
+			return nil, err
+		}
+
+		files, err := g.repos.GetFiles(context.TODO(), git.RepoURL, git.Revision, templatedPath, noRevisionCache, verifyCommit)
+		if err != nil {
+			return nil, err
+		}
+		for extraFilePath, content := range files {
+			extraParamFilesContent[extraFilePath] = content
+		}
+		// Extract the unduplicated map into a list, and sort by path to ensure a deterministic
+		// processing order in the subsequent step
+		allPaths := []string{}
+		for path := range files {
+			allPaths = append(allPaths, path)
+		}
+		sort.Strings(allPaths)
```

Comment:
This might lead to unexpected behavior, since the order matters for how things get combined IIUC. People would expect the original order.

Reply:
Here it's handling a glob, not the list of provided files.
If we decide to support globs there, we absolutely need to sort the files matching to be deterministic.

Reply:
Ah okay